src/core/border_router/router_advertisement.cpp

```diff
@@ -113,62 +86,85 @@ void PrefixInfoOption::SetPrefix(const Ip6::Prefix &aPrefix)
+           (GetPreferredLifetime() <= GetValidLifetime());
+}
 
-    mPrefix.Clear();
+//----------------------------------------------------------------------------------------------------------------------
+// RouteInfoOption
+
+void RouteInfoOption::Init(void)
+{
+    Clear();
+    SetType(Type::kRouteInfo);
 }
 
 void RouteInfoOption::SetPreference(RoutePreference aPreference)
 {
-    mReserved &= ~kPreferenceMask;
-    mReserved |= (NetworkData::RoutePreferenceToValue(aPreference) << kPreferenceOffset) & kPreferenceMask;
+    mResvdPrf &= ~kPreferenceMask;
+    mResvdPrf |= (NetworkData::RoutePreferenceToValue(aPreference) << kPreferenceOffset) & kPreferenceMask;
 }
 
 RouteInfoOption::RoutePreference RouteInfoOption::GetPreference(void) const
 {
-    return NetworkData::RoutePreferenceFromValue((mReserved & kPreferenceMask) >> kPreferenceOffset);
+    return NetworkData::RoutePreferenceFromValue((mResvdPrf & kPreferenceMask) >> kPreferenceOffset);
 }
 
 void RouteInfoOption::SetPrefix(const Ip6::Prefix &aPrefix)
 {
-    // The total length (in bytes) of a Router Information Option
-    // is: (8 bytes fixed option header) + (0, 8, or 16 bytes prefix).
-    // Because the length of the option must be padded with 8 bytes,
-    // the length of the prefix (in bits) must be padded with 64 bits.
-    SetLength((aPrefix.mLength + kLengthUnit * CHAR_BIT - 1) / (kLengthUnit * CHAR_BIT) + 1);
-
+    SetLength(OptionLengthForPrefix(aPrefix.mLength));
     mPrefixLength = aPrefix.mLength;
-    mPrefix       = AsCoreType(&aPrefix.mPrefix);
+    memcpy(GetPrefixBytes(), aPrefix.GetBytes(), aPrefix.GetBytesSize());
 }
 
-Ip6::Prefix RouteInfoOption::GetPrefix(void) const
+void RouteInfoOption::GetPrefix(Ip6::Prefix &aPrefix) const
 {
-    Ip6::Prefix prefix;
-
-    prefix.Set(mPrefix.GetBytes(), mPrefixLength);
-    return prefix;
+    aPrefix.Set(GetPrefixBytes(), mPrefixLength);
 }
 
 bool RouteInfoOption::IsValid(void) const
 {
-    return (GetLength() == 1 || GetLength() == 2 || GetLength() == 3) &&
-           (mPrefixLength <= OT_IP6_ADDRESS_SIZE * CHAR_BIT) && NetworkData::IsRoutePreferenceValid(GetPreference());
+    return (GetSize() >= kMinSize) && (mPrefixLength <= Ip6::Prefix::kMaxLength) &&
+           (GetLength() >= OptionLengthForPrefix(mPrefixLength)) &&
+           NetworkData::IsRoutePreferenceValid(GetPreference());
+}
+
+uint8_t RouteInfoOption::OptionLengthForPrefix(uint8_t aPrefixLength)
+{
+    static constexpr uint8_t kMaxPrefixLenForOptionLen1 = 0;
+    static constexpr uint8_t kMaxPrefixLenForOptionLen2 = 64;
+
+    uint8_t length = 3;
+
+    // The Option Length can be 1, 2, or 3 depending on the prefix
+    // length
+    //
+    // - 1 when prefix len is zero.
+    // - 2 when prefix len is less then or equal to 64.
+    // - 3 otherwise.
+
+    if (aPrefixLength == kMaxPrefixLenForOptionLen1)
+    {
+        length = 1;
+    }
+    else if (aPrefixLength <= kMaxPrefixLenForOptionLen2)
+    {
+        length = 2;
+    }
```

Comment:
nit: just for aligning with the comment :)
```suggestion
    uint8_t length;

    // The Option Length can be 1, 2, or 3 depending on the prefix
    // length
    //
    // - 1 when prefix len is zero.
    // - 2 when prefix len is less then or equal to 64.
    // - 3 otherwise.

    if (aPrefixLength == kMaxPrefixLenForOptionLen1)
    {
        length = 1;
    }
    else if (aPrefixLength <= kMaxPrefixLenForOptionLen2)
    {
        length = 2;
    }
    else
    {
        length = 3;
    }
```